spec/message.ebnf

```diff
@@ -44,13 +39,13 @@ NameChar ::= NameStart | [0-9] | "-" | "." | #xB7
            | [#x0300-#x036F] | [#x203F-#x2040]
 
 /* Literals */
-Literal ::= '(' (LiteralChar | LiteralEscape)* ')' /* ws: explicit */
+Literal ::= '(' (LiteralChar | LiteralEscape)* ')'
 LiteralChar ::= AnyChar - ('(' | ')' | Esc)
 
 /* Escape sequences */
 Esc ::= '\'
 TextEscape ::= Esc Esc | Esc '{' | Esc '}'
 LiteralEscape ::= Esc Esc | Esc '(' | Esc ')'
 
-/* WhiteSpace */
-WhiteSpace ::= #x9 | #xD | #xA | #x20 /* ws: definition */
+/* White space */
+s ::= (#x9 | #xD | #xA | #x20)+
```

Comment:
Use `LWSP`, which is built-in, instead of defining our own?

Reply:
Looking at the definition in https://datatracker.ietf.org/doc/html/rfc5234#appendix-B, `LWSP` excludes standalone `LF`, which I don't think we want to do?

Reply:
Also, the same appendix says that it _contains some basic rules that are in common use_ and that _[b]asic rules are in uppercase_. Does _common use_ mean that they're built-in, or perhaps that they're expected to be defined in a certain way?

The reason I'm asking is that I found an ABNF parser at https://author-tools.ietf.org/abnf and it complained about `DIGIT` and `ALPHA` being undefined. Bug or intended behavior?

Reply:
I would prefer keeping an explicit whitespace definition within the spec, rather than introducing an external dependency.

Reply:
> Also, the same appendix says that it _contains some basic rules that are in common use_ and that _[b]asic rules are in uppercase_. Does _common use_ mean that they're built-in, or perhaps that they're expected to be defined in a certain way?

I think in an RFC, you would usually say that the following rules are imported from RFC 5234, and give a list of the rules you need.

> The reason I'm asking is that I found an ABNF parser at https://author-tools.ietf.org/abnf and it complained about `DIGIT` and `ALPHA` being undefined. Bug or intended behavior?

DIGIT and ALPHA are too frequent for such a bug not having been detected. I'd guess you have to add the rules you reference from RFC 5234 by hand.

Reply:
Bill's parser (which powers author-tools) has always complained about the built-ins. I think most I-D authors ignore the warnings (I know I learned to). For example, if you run BCP47's ABNF in the tool, it complains about all of the built-ins (and BCP47 is a published set of RFCs blessed by the editor):

```abnf
Language-Tag = langtag / privateuse / grandfathered
langtag = language [ "-" script ] [ "-" region ] *( "-" variant ) *( "-" extension ) [ "-" privateuse ]
; ALPHA UNDEFINED
language = ( 2*3ALPHA [ "-" extlang ] ) / 4ALPHA / 5*8ALPHA
extlang = 3ALPHA *2( "-" 3ALPHA )
script = 4ALPHA
; DIGIT UNDEFINED
region = 2ALPHA / 3DIGIT
variant = 5*8alphanum / ( DIGIT 3alphanum )
extension = singleton 1*( "-" ( 2*8alphanum ) )
singleton = DIGIT / %x41-57 / %x59-5A / %x61-77 / %x79-7A
privateuse = "x" 1*( "-" ( 1*8alphanum ) )
grandfathered = irregular / regular
irregular = "en-GB-oed" / "i-ami" / "i-bnn" / "i-default" / "i-enochian" / "i-hak" / "i-klingon" / "i-lux" / "i-mingo" / "i-navajo" / "i-pwn" / "i-tao" / "i-tay" / "i-tsu" / "sgn-BE-FR" / "sgn-BE-NL" / "sgn-CH-DE"
regular = "art-lojban" / "cel-gaulish" / "no-bok" / "no-nyn" / "zh-guoyu" / "zh-hakka" / "zh-min" / "zh-min-nan" / "zh-xiang"
alphanum = ( ALPHA / DIGIT )
obs-language-tag = primary-subtag *( "-" subtag )
primary-subtag = 1*8ALPHA
subtag = 1*8( ALPHA / DIGIT )
; CRLF UNDEFINED
registry = record *( "%%" CRLF record )
record = 1*field
field = ( field-name field-sep field-body CRLF )
field-name = ( ALPHA / DIGIT ) [ *( ALPHA / DIGIT / "-" ) ( ALPHA / DIGIT ) ]
; SP UNDEFINED
field-sep = *SP ":" *SP
```

Reply:
> In addition to +1 on @eemeli's [comment](https://github.com/unicode-org/message-format-wg/pull/344/files#r1105089539) to allow for whitespace between `Declaration`s (the `let ...`) [...]

Did you mean to say "to allow" or "to require"? @eemeli's comment uses a stronger language.

> Same applies to the [comment](https://github.com/unicode-org/message-format-wg/pull/344/files#r1105091618) for whitespace between each variant in a selection message. At least, by "natural", I mean that I think this is how we have already written all of our example messages in our documentation so far.

I agree about the "natural". I think we should recommend readable formatting and lead by example. This PR is more about defining the bounds of what's considered valid, even if it's formatted horribly.

Reply:
@stasm Ah, yep, I meant "require".